pyam/core.py

```diff
@@ -261,35 +261,57 @@ def metadata(self, meta, name=None):
             column to be added to metadata
             (by `['model', 'scenario']` index if possible)
         name: str
-            category column name (if not given by meta pd.Series.name)
+            meta column name (if not given by meta pd.Series.name)
+        index: pyam.IamDataFrame or pd.MultiIndex, optional
+            index to be used for setting meta column
         """
-        if isinstance(meta, pd.Series) and \
-                set(['model', 'scenario']).issubset(meta.index.names):
-            meta.name = name or meta.name
-            # reduce index dimensions to model-scenario only
-            meta = (meta
-                    .reset_index()
-                    .loc[:, ['model', 'scenario', meta.name]]
-                    .set_index(['model', 'scenario'])
-                    )
-            # raise error if index is not unique
-            if meta.index.duplicated().any():
-                raise ValueError("non-unique ['model', 'scenario'] index!")
-            # check if trying to add model-scenario index not existing in self
-            diff = meta.index.difference(self.meta.index)
-            if not diff.empty:
-                error = "adding metadata for non-existing scenarios '{}'!"
-                raise ValueError(error.format(diff))
-            self.meta = meta.combine_first(self.meta)
-            #  quickfix for pandas.combine_first(), issue #7509
-            self.meta['exclude'] = self.meta['exclude'].astype('bool')
+        if not name and not hasattr(meta, 'name'):
+            raise ValueError('Must pass a name or use a pd.Series')
+
+        # use meta.index if index arg is an IamDataFrame
+        if index is not None and isinstance(index, IamDataFrame):
+            index = index.meta.index
+
+        # create pd.Series from index and meta if provided
+        _meta = meta if index is None else pd.Series(data=meta, index=index,
+                                                     name=name)
+
+        try:
+            append_by_idx = set(META_IDX).issubset(_meta.index.names)
+        except AttributeError:
+            append_by_idx = False
+
+        # if not possible to append by index
+        if not append_by_idx:
+            if isinstance(meta, collections.Iterable) and not isstr(meta):
```

Comment:
Not blocking, but I recently implemented `isscalar` in message_ix: https://github.com/iiasa/message_ix/blob/master/message_ix/utils.py#L29